src/astro/sql/operators/sql_decorator_refactor.py

```diff
@@ -0,0 +1,167 @@
+        role: Optional[str] = None,
+        raw_sql=False,
+        sql="",
+        **kwargs,
+    ):
+        self.kwargs = kwargs or {}
+        self.op_kwargs: Dict = self.kwargs.get("op_kwargs") or {}
+        self.output_table: Optional[Table] = self.op_kwargs.pop("output_table", None)
+        self.handler = self.op_kwargs.pop("handler", None)
+        self.sql = sql
+
+        super().__init__(
+            sql=sql,
+            autocommit=autocommit,
+            handler=handler,
+            parameters=parameters,
+            conn_id=self.op_kwargs.pop("conn_id", conn_id),
+            database=self.op_kwargs.pop("database", database),
+            schema=self.op_kwargs.pop("schema", schema),
+            warehouse=self.op_kwargs.pop("warehouse", warehouse),
+            role=self.op_kwargs.pop("role", role),
+            raw_sql=raw_sql,
+            **kwargs,
+        )
+
+    @staticmethod
+    def convert_old_table_to_new(table: OldTable) -> Table:
+        """
+        This function is only temporary until other functions use the new table format.
+
+        Converts a TempTable or a Table object into the new Table format.
+        :param table: old table
+        """
+        if isinstance(table, TempTable):
+            table = table.to_table("")
+        if isinstance(table, OldTable):
+            table = Table(
+                conn_id=table.conn_id,
+                name=table.table_name,
+                metadata=Metadata(
+                    schema=table.schema,
+                    warehouse=table.warehouse,
+                    database=table.database,
+                ),
+            )
+        return table  # type: ignore
+
+    def handle_table_conversions(self):
+        """This is a temporary holdover until all other functions use the new table format"""
+        self.op_args = tuple(
+            self.convert_old_table_to_new(t) if isinstance(t, OldTable) else t
+            for t in self.op_args  # type: ignore
+        )  # type: ignore
+        self.op_kwargs = {
+            k: self.convert_old_table_to_new(t) if isinstance(t, OldTable) else t
+            for k, t in self.op_kwargs.items()
+        }
+
+    def execute(self, context: Dict):
+        self.handle_table_conversions()
+
+        self._set_variables_from_first_table()
+        self.database_impl = create_database(self.conn_id)
+
+        # Find and load dataframes from op_arg and op_kwarg into Table
+        self.create_output_table(self.output_table_name)
+        self.load_op_arg_dataframes_into_sql()
+        self.load_op_kwarg_dataframes_into_sql()
+
+        # Get SQL from function and render templates in the SQL String
+        self.read_sql_from_function()
+        self.move_function_params_into_sql_params(context)
+        self.template(context)
+        self.parameters = self.database_impl.process_sql_parameters(self.parameters)  # type: ignore
+        self.handle_schema()
+
+        if self.raw_sql:
```

Comment:
It feels like we are giving two very different behaviors to a single decorator - would it make sense splitting them?

**Behavior 1**

Input:
- table(s)
- SQL statement

Output:
- table

**Behavior 2**

Input:
- table(s)
- SQL Statement
- function

Output:
- the result of running SQL statement + applying function (this can potentially trash the xcom database depending on the backend used)

Reply:
I think we could split in the future, but for now 90% of the functionality is the same (Everything leading up to actually running the query).